File: scripts/test_retrieval_parity.py
"""导出 Python 端 BM25 top-10，供浏览器端实现做逐条对比。"""

import json
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from retrieve import Retriever  # noqa: E402

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

QUERIES = [
    "贵州茅台2025年营业收入是多少",
    "五粮液的毛利率",
    "合同负债",
    "2026年上半年归母净利润",
    "存货跌价准备",
    "经销商数量",
    "省外市场收入占比",
    "每10股派发现金红利",
    "销售费用率",
    "经营活动产生的现金流量净额",
]


def main():
    r = Retriever()
    out = []
    for q in QUERIES:
        scores = r.bm25(q)
        top = sorted(scores, key=lambda i: (-scores[i], i))[:10]
        out.append({"query": q, "python_top": top})
    path = os.path.join(ROOT, "data", "out", "parity_ref.json")
    json.dump(out, open(path, "w", encoding="utf-8"), ensure_ascii=False, indent=1)
    print("wrote", path)


if __name__ == "__main__":
    sys.exit(main())

File: scripts/test_retrieval_parity.mjs
/* 校验浏览器端 BM25 实现（web/js/bm25.js）与离线 Python 版检索结果一致 */
import { readFileSync } from "node:fs";
import { gunzipSync } from "node:zlib";
import { BM25Index } from "../web/js/bm25.js";
import { expandQuery } from "../web/js/rag.js";

const root = new URL("../", import.meta.url).pathname;
const bmBytes = gunzipSync(readFileSync(root + "web/data/bm25.bin.gz"));
const buf = bmBytes.buffer.slice(bmBytes.byteOffset, bmBytes.byteOffset + bmBytes.byteLength);
const bm = new BM25Index(buf);
const lens = new Uint32Array(readFileSync(root + "web/data/doclen.bin").buffer.slice(0));
const ref = JSON.parse(readFileSync(root + "data/out/parity_ref.json", "utf8"));

let bad = 0;
for (const { query, python_top } of ref) {
  const scores = bm.score(expandQuery(query), null, lens);
  const jsTop = [...scores.entries()]
    .sort((a, b) => b[1] - a[1] || a[0] - b[0])
    .slice(0, 10)
    .map(([i]) => i);
  const same = jsTop.length === python_top.length && jsTop.every((v, i) => v === python_top[i]);
  if (!same) {
    bad++;
    console.log("MISMATCH:", query);
    console.log("  py:", python_top.join(","));
    console.log("  js:", jsTop.join(","));
  }
}
console.log(bad === 0 ? `BM25 排序完全一致（${ref.length} 条查询）` : `${bad}/${ref.length} 条不一致`);
